### ISSUE TITLE
Precedence of Java Toolchains

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#16194]
This request / report consists of two parts

1. Please document the exact precedence about which Toolchain is used
2. Please change the precedence

Regarding point 2, I'd usually expect a locally configured value to win over some system-wide or auto-used value or similar.
But with Java Toolchains this is not the case.
I thought if `org.gradle.java.installations.fromEnv` is configured and a matching toolchain is found, this should be the one that is used, as it is what the user configured locally.
But this is not the case. If there is an auto-provisioned toolchain, then that is used (or maybe because it is newer than the configured one, I don't know because of point 1.)

This means, if you have no provisioned toolchain that matches and run build A where toolchain X is configured via environment variable, then X is used.
Then you run build B where no environment variables are configured, so toolchain Y is auto-provisioned.
If you now run build A again, it is suddenly built with toolchain Y, even with X being configured via environment variable.

Of course I could disable auto-provisioning, but that is not intended, because it should also work out-of-the-box for people where the environment variable is not set. But if it is set, it should take precedence.

If you don't like to change the current default logic, whatever it is, then maybe an additional property would make sense with which you then could configure the precedence or similar.

**Comment by ljacomet** [comment]
Thanks for the report. I agree that we the way a toolchain is resolved should be documented.
However I disagree with your request for changing it. We cannot include the "source" of the toolchain in the ordering, that would make any changes to the set of available sources problematic.

I am assuming that in your case you had a version of Java X, then downloaded a more recent release of version X and then that one wins.
I believe that's the right behavior and if the exact revision of version X matters, we could make that a selection criteria. The assumption though was that it only mattered in specific cases and you could then disable auto provisioning for those.

**Comment by Vampire** [comment]
> We cannot include the "source" of the toolchain in the ordering, that would make any changes to the set of available sources problematic.

What problems do you see there? 

> I am assuming that in your case you had a version of Java X, then downloaded a more recent release of version X and then that one wins.

Exactly

> I believe that's the right behavior and if the exact revision of version X matters, we could make that a selection criteria.

This could also be the case, but wasn't in my use-case.

> The assumption though was that it only mattered in specific cases and you could then disable auto provisioning for those.

Unfortunately not, the use-case is, if someone configures an explicit version using an environment variable, for example `JAVA_HOME` or `JAVA11_HOME`, then the user selection should win, that way you can for example easily test with different minor versions. But if no environment variable is set, or not to a compatible version, then one should be auto-provisioned.

That's like local Gradle properties that win over checked in Gradle properties.

**Comment by smoothreggae** [comment]
I second the request for the sequence/behavior to be documented. In my specific use case, I have a build that specifies a toolchain of Java 8. It is possible that Java 11 could be used to run this build, so the toolchain was very useful to ensure that I don't get class files that are incompatible. When I had `JAVA_HOME` pointing to a local installation of Java `8u275` and ran the build initially, my local installation was used. When I ran the build with Java 11, Gradle (via the toolchain) downloaded an AdoptOpenJDK release of Java `8u272`. I upgraded my local installation of Java to `8u282`. With `JAVA_HOME` pointing to this location, I ran the build, but now Gradle refuses to use the compatible version of Java that I am running the build with and insists on using the AdoptOpenJDK `8u272` it had downloaded earlier.

I may be misunderstanding the [documentation for toolchains](https://docs.gradle.org/current/userguide/toolchains.html), but it does seem to suggest that Gradle should pick up my local installation if it's valid and not insist on using a Java 8 downloaded via the toolchain. 

Here's what it says under [Consuming Toolchains](https://docs.gradle.org/current/userguide/toolchains.html#sec:consuming):

> Gradle detects locally installed JVMs
> Gradle chooses a JRE/JDK matching the requirements of the build
> If no matching JVM is found, it will automatically download a matching JDK from AdoptOpenJDK

Later, under [Auto Provisioning](https://docs.gradle.org/current/userguide/toolchains.html#sec:provisioning):

> If Gradle can’t find a locally available toolchain which matches the requirements of the build, it will automatically try to download it from AdoptOpenJDK. By default, it will request a HotSpot JDK matching the current operating system and architecture.

What Gradle is doing in my case seems to suggest that JDKs downloaded via toolchains have precedence over local installations (including the one being used to run the build).
